### DISCUSSION TITLE
Adventure map, optimization of the hero's predefined path

### DISCUSSION THREAD

**Discussion by LeHerosInconnu** [ihhub/fheroes2#6106] [General]
"In the original game, the hero's predefined path favors horizontal and vertical moves first, rather than diagonal moves, which allows the hero to benefit from the bonus given to the last move (if the hero can move horizontally or vertically then he can move diagonally).

In fheroes2, this is not the case, and it penalizes the hero by reducing his movement."

Here is an example using the predefined path first, then "modifying" the predefined path.
In the first case, the fourth hero cannot be attacked, which is possible in the second case.

In fheroes2:

Using the predefined path:

https://user-images.githubusercontent.com/43583994/200172373-6c596d6b-e1bd-4a8c-98ea-98ceee1f518b.mp4

"Modifying" the predefined path:

https://user-images.githubusercontent.com/43583994/200172393-005f2161-bb8f-4667-8bbc-c9d8aac65888.mp4

Here is the save file:
[Before auto path 01.zip](https://github.com/ihhub/fheroes2/files/9945894/Before.auto.path.01.zip)

Related closed issue: https://github.com/ihhub/fheroes2/issues/4510.

**Comment by oleg-derevenetz** [comment]
I have an impression that this was already discussed. This "optimization" is no more than an ad-hoc illusion, and pathfinding should in fact foresee the future, because if the last hero would be in a slightly different place, the first path could be more optimal:

https://user-images.githubusercontent.com/32623900/200231497-c6fadfeb-2ef8-44aa-ab3f-5a4c11b3e422.mp4

vs

https://user-images.githubusercontent.com/32623900/200231337-1184835d-95ad-463b-8900-b190977d6107.mp4

That's the player's task, not the pathfinder's - the player knows what his next move should be, but pathfinder doesn't. I propose to close the topic of "the pathfinder that should foresee the future" :)

**Reply by LeHerosInconnu** [reply]
Hello @oleg-derevenetz,

> I have an impression that this was already discussed. This "optimization" is no more than an ad-hoc illusion, and pathfinding should in fact foresee the future, because if the last hero would be in a slightly different place, the first path could be more optimal:
> 
>  2022-11-07.08.09.43.mp4 
> vs
> 
>  2022-11-07.08.10.08.mp4 
> That's the player's task, not the pathfinder's - the player knows what his next move should be, but pathfinder doesn't. I propose to close the topic of "the pathfinder that should foresee the future" :)

It's not about predicting the future, even with the help of "Precogs". :rofl: 

But to favor horizontal and vertical movement first, rather than diagonal moves for the hero's predefined path.
As it is the case in the original game.

In the original game, the predefined path always starts by taking into account the vertical and horizontal moves (each step of the hero) first (when possible).
In fheroes2, this is not the case, and we end up with a different predefined path depending on where we click.

The same rule of operation as the original game should be applied in fheroes2.

In original game:

https://user-images.githubusercontent.com/43583994/200609676-e5f156fe-6bc5-437d-ad8c-a64ff909b732.mp4

In fheroes2:

https://user-images.githubusercontent.com/43583994/200611269-a6f88d3f-5e2c-49ee-a1f3-7fc50e689347.mp4

Here is the test scenario file;
[0 9927.zip](https://github.com/ihhub/fheroes2/files/9962876/0.9927.zip)

**Comment by oleg-derevenetz** [comment]
Hi @LeHerosInconnu

> But to favor horizontal and vertical movement first, rather than diagonal moves for the hero's predefined path.
> As it is the case in the original game.

This topic tends to pretend that this behavior gives the player some advantages that are not really there, and I hope I have demonstrated this convincingly.

**Comment by LeHerosInconnu** [comment]
@oleg-derevenetz.

> Hi @LeHerosInconnu
> 
> > But to favor horizontal and vertical movement first, rather than diagonal moves for the hero's predefined path.
> > As it is the case in the original game.
> 
> This topic tends to pretend that this behavior gives the player some advantages that are not really there, and I hope I have demonstrated this convincingly.

No, there is a difference.
And it's a loss of movement for the hero.

In original game (from video above):
![vlcsnap-2022-11-08-18h22m38s979](https://user-images.githubusercontent.com/43583994/200634616-a588909a-c87d-4170-b145-7d54046c4f7c.png)

In fheroes2 (from video above):
![vlcsnap-2022-11-08-18h24m06s773](https://user-images.githubusercontent.com/43583994/200634693-a9c128ff-055b-420a-94c3-f90a57ca975e.png)

**Comment by oleg-derevenetz** [comment]
@LeHerosInconnu

> No, there is a difference.
> And it's a loss of movement for the hero.

There is no loss of movement if this path will be laid directly to the end point, without all these artificial intermediate routes. I remember there was also an argument "what if hero will be stopped while moving to the endpoint", and it was also answered - in general, this also requires some psychic powers to predict the future. There was even an [example of two paths](https://github.com/ihhub/fheroes2/pull/4626#issuecomment-974709710), one of them was "better" in one case, and the second one was "better" in another case - depending on the exact point when hero will be re-routed.

**Reply by LeHerosInconnu** [reply]
Hello @oleg-derevenetz,

> @LeHerosInconnu
> 
> > No, there is a difference.
> > And it's a loss of movement for the hero.
> 
> There is no loss of movement if this path will be laid directly to the end point, without all these artificial intermediate routes. I remember there was also an argument "what if hero will be stopped while moving to the endpoint", and it was also answered - in general, this also requires some psychic powers to predict the future. There was even an [example of two paths](https://github.com/ihhub/fheroes2/pull/4626#issuecomment-974709710), one of them was "better" in one case, and the second one was "better" in another case - depending on the exact point when hero will be re-routed.

It can happen during the game that the hero's movement is stopped by the player before reaching the end of the predefined hero path.
For example, with a hero who reveals the terrain under the fog of war and discovers a point of interest (an opposing hero, a troop of wandering creatures, a resource generator, etc.).
This is what I wanted to show with "all these artificial intermediate routes".
If the hero spends 150 movement points with his first step diagonally, the extra 50 points spent compared to the 100 points spent with a horizontal or vertical first step cannot be recovered later in the hero's movement.

**Comment by LeHerosInconnu** [comment]
Another example.

It's day 7.
The player has the opportunity to pick up a pile of ore, which will enable him to build the Green Tower, with 4 green dragons available for recruiting the following week.
An important advantage for the rest of the scenario.
Will he succeed?

Let's try with Heroes 2 first:

https://github.com/ihhub/fheroes2/assets/43583994/7c303bb0-c276-48c3-b2e6-4a4eb6cb3e34

**It's a success!!!**

Next, let's try fheroes2:

https://github.com/ihhub/fheroes2/assets/43583994/ec141f41-d13d-4c6a-8ae5-3dfc8f18ec57

**Total failure!!!**

The 50 extra hero movement points spent by the diagonal movement at the start of the hero's path to attack the peasant are permanently lost for the rest of the hero's movement, and will never be recovered.

This doesn't happen in the original game, because the game first favors horizontal and vertical movements for heroes, which consume fewer movement points.

Here is the scenario test and the save files:
[Dragon or not Dragon 01.zip](https://github.com/ihhub/fheroes2/files/11593610/Dragon.or.not.Dragon.01.zip)

**Comment by LeHerosInconnu** [comment]
This is planned in my mod: https://github.com/ihhub/fheroes2/discussions/10109.
I close the discussion.
